pipeline/pipeline.go

```diff
@@ -0,0 +1,182 @@
+package pipeline
+
+import (
+	"bytes"
+	"encoding/csv"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/fullstorydev/hauser/config"
+	"github.com/nishanths/fullstory"
+)
+
+// A Pipeline downloads data exports from fullstory.com and saves them to local disk
+type Pipeline struct {
+	transformRecord func(map[string]interface{}) ([]string, error)
+	metaTime        time.Time
+	metaCh          chan fullstory.ExportMeta
+	expCh           chan ExportData
+	saveCh          chan SavedExport
+	recsCh          chan RecordGroup
+	errCh           chan error
+	conf            *config.Config
+}
+
+// NewPipeline returns a new Pipeline base on the configuration provided
+func NewPipeline(conf *config.Config, transformRecord func(map[string]interface{}) ([]string, error)) *Pipeline {
+	return &Pipeline{
+		conf:            conf,
+		transformRecord: transformRecord,
+		metaCh:          make(chan fullstory.ExportMeta),
+		expCh:           make(chan ExportData),
+		recsCh:          make(chan RecordGroup),
+		saveCh:          make(chan SavedExport),
+		errCh:           make(chan error),
+	}
+}
+
+// Start begins pipeline downloading and processing bundles after the provided time. This function returns a channel
+// that will contain the filenames to which the data was saved and a channel that contains any errors that occur in the
+// pipeline. These must be retrieved from the channel to continue processing.
+func (p *Pipeline) Start(startTime time.Time) (chan SavedExport, chan error) {
+	p.metaTime = startTime
+	go p.metaFetcher()
+	go p.exportFetcher()
+	go p.recordGrouper()
+	go p.recordsSaver()
+	return p.saveCh, p.errCh
+}
+
+// Stop is used to stop the pipeline from processing any more exports
+func (p *Pipeline) Stop() {
+	close(p.expCh)
+	close(p.metaCh)
+	close(p.recsCh)
+	close(p.saveCh)
+	close(p.errCh)
+}
+
+func (p *Pipeline) metaFetcher() {
+	for {
+		fs := getFSClient(p.conf)
+		exportList, err := fs.ExportList(p.metaTime)
+		if err != nil {
+			p.errCh <- fmt.Errorf("could not fetch export list: %s", err)
+			return
+		}
+		for _, meta := range exportList {
+			p.metaCh <- meta
+			p.metaTime = meta.Stop
+		}
+		if len(exportList) == 0 {
+			log.Printf("No exports pending; sleeping %s", p.conf.CheckInterval.Duration)
+			time.Sleep(p.conf.CheckInterval.Duration)
+		}
+	}
+}
+
+func (p *Pipeline) exportFetcher() {
+	for meta := range p.metaCh {
+		fs := getFSClient(p.conf)
+		data, err := getDataWithRetry(fs, meta)
+		if err != nil {
+			p.errCh <- fmt.Errorf("error fetching data export: %s", err)
+			return
+		}
+		p.expCh <- data
+	}
+}
+
+func (p *Pipeline) recordGrouper() {
+	var recs []Record
```

Comment:
It won't be workable to hold all the records in memory here. We're talking tens or hundreds of megabytes, or for extremely large clients a gigabyte or more of data in a single bundle. This has to work on a model that doesn't hold all the records in memory, hence the original design that just streamed data to disk.